fix(treesitter): update window if inspected buffer is still open

guilhas07 opened the pull request:
Problem:
1 - When interacting with multiple InspectTree and the source buffer
windows there was a high chance of errors due to the window ids not
being updated and validated.

2 - Not all InspectTree windows were closed when
the source buffer was closed.

Solution:
1- Update InspectTree window id on `CursorMoved` event
and validate source buffer window id before trying to navigate to
it.
2- Close all InspectTree windows

Let me know if there is something that I need to change!

justinmk commented on runtime/lua/vim/treesitter/dev.lua:
```diff
@@ -501,12 +501,21 @@ function M.inspect_tree(opts)
+        close_win(w)
+        return true
+      end
+      for _, new_win in pairs(windows) do
+        if new_win ~= win then
+          win = new_win
+          break
+        end
+      end
```
finding an item is a good use-case for vim.iter

lewis6991 replied:
I would do `vim.iter(pairs(windows))`. ArrayIter doesn't really help with find AFAIK

guilhas07 replied:
> I would do `vim.iter(pairs(windows))`. ArrayIter doesn't really help with find AFAIK

Done. Despite not really knowing the difference 😅. Would appreciate it if you could explain it or point me somewhere though because I could only find mentions of `vim.iter(ipairs(...))` in the docs.

justinmk commented on runtime/lua/vim/treesitter/dev.lua:
```diff
@@ -501,12 +501,19 @@ function M.inspect_tree(opts)
+        return true
+      end
+
+      win = vim.iter(windows):find(function(new_win)
```
This is somewhat subtle (updating shared state that is "far" from this scope). So worth a comment.
```suggestion
      -- Maintain a valid window, if the original window was closed.
      win = vim.iter(windows):find(function(new_win)
```

justinmk commented on runtime/lua/vim/treesitter/dev.lua:
```diff
@@ -501,12 +501,19 @@ function M.inspect_tree(opts)
     callback = function()
-      close_win(w)
+      local windows = vim.fn.win_findbuf(buf) --[=[@as integer[]]=]
+      if #windows == 1 then
+        close_win(w)
+        return true
```
This seems weird. Why is the WinClosed handler closing the window? Seems redundant.

Also why is the window being stored in both `w` and `win`? https://github.com/justinmk/neovim/blob/52d2851ca4e747146c30e1996c08bf504d99fe95/runtime/lua/vim/treesitter/dev.lua#L329-L340

What is the difference between `w` and `win` ?

justinmk replied:
> * `win` and `buf` for the buffer that has the code
> * `w` and `b` for the InspectTree buffer

for now let's add docstrings on those. can use that exact wording

justinmk commented on runtime/lua/vim/treesitter/dev.lua:
```diff
@@ -501,12 +506,50 @@ function M.inspect_tree(opts)
+    end,
+  })
+
+  api.nvim_create_autocmd({ 'WinClosed' }, {
```
wait where did this extra handler come from?

this PR is turning into a lot of code, for a pretty minor use-case.

is there a simpler approach we can use? can we enforce a constraint to make things simpler?

guilhas07 replied:
> wait where did this extra handler come from?

I observed a similar situation that happened if the user opened two InspectTree buffers and closed the original one. Similar to the `BufLeave` events, I think this needs to be duplicated.

>this PR is turning into a lot of code, for a pretty minor use-case.

Yeah, I get it. But it can be really annoying because it forces a user to restart all their window setup. And the repetitive nature of the code is the true reason for so many lines.

>is there a simpler approach we can use? can we enforce a constraint to make things simpler?

The only other way I can think of doing this would be to always validate the windows on the events in which they are used, but that would decrease performance. Do you think this approach could be better?

justinmk commented on test/functional/treesitter/inspect_tree_spec.lua:
```diff
@@ -114,4 +114,56 @@ describe('vim.treesitter.inspect_tree', function()
+    eq('', n.api.nvim_get_vvar('errmsg'))
+  end)
+
+  it('updates inspect tree window', function()
```
would it save a lot of code if this test is merged into the one above?

guilhas07 replied:
Only managed to reduce 9 lines.

justinmk commented on runtime/lua/vim/treesitter/dev.lua:
```diff
@@ -506,7 +523,10 @@ function M.inspect_tree(opts)
     callback = function()
-      close_win(w)
+      -- close all tree windows
+      vim.iter(pairs(vim.fn.win_findbuf(b))):each(function(_, window)
+        close_win(window)
```
nit: since we are no longer doing find() , or modifying the iterable in other ways, each() is actually more verbose than just using pairs().

```suggestion
      for _, window in pairs(vim.fn.win_findbuf(b)) do
        close_win(window)
```

justinmk reviewed (approved):
Thanks for finding a reasonably simple approach

guilhas07 committed:
fix(treesitter): update and close source and inspector windows correctly

Problem:
1 - When interacting with multiple InspectTree and the source buffer
windows there was a high chance of errors due to the window ids not
being updated and validated.

2 - Not all InspectTree windows were closed when
the source buffer was closed.

Solution:
1- Update InspectTree window id on `CursorMoved` event
and validate source buffer window id before trying to navigate to
it.
2- Close all InspectTree windows

guilhas07 commented:
> Thanks for finding a reasonably simple approach

Thank you for guiding me to it!
